server/src/main/java/org/elasticsearch/search/profile/SearchProfileResults.java

```diff
@@ -173,4 +199,53 @@ private static void parseProfileResultsEntry(XContentParser parser, Map<String,
         result.getQueryPhase().setSearchProfileDfsPhaseResult(searchProfileDfsPhaseResult);
         searchProfileResults.put(id, result);
     }
+
+    /**
+     * Parsed representation of a composite id used for shards in a profile.
+     * The composite id format is specified/created via the {@code SearchShardTarget} method.
+     * @param nodeId nodeId that the shard is on
+     * @param indexName index being profiled
+     * @param shardId shard id being profiled
+     * @param clusterName if a CCS search, the remote clusters will have a name in the id. Local clusters will be null.
+     */
+    record ShardProfileId(String nodeId, String indexName, int shardId, @Nullable String clusterName) {}
+
+    /**
+     * Parse the composite "shard id" from the profiles output, which comes from the
+     * {@code SearchShardTarget.toString()} method, into its separate components.
+     * <p>
+     * One of two expected patterns is accepted:
+     * <p>
+     * 1) [nodeId][indexName][shardId]
+     * example: [2m7SW9oIRrirdrwirM1mwQ][blogs][1]
+     * <p>
+     * 2) [nodeId][clusterName:indexName][shardId]
+     * example: [UngEVXTBQL-7w5j_tftGAQ][remote1:blogs][0]
+     *
+     * @param compositeId see above for accepted formats
+     * @return ShardProfileId with parsed components or null if the compositeId has an unsupported format
+     */
+    static ShardProfileId parseProfileShardId(String compositeId) {
+        if (Strings.isNullOrEmpty(compositeId)) {
+            return null;
+        }
+        Pattern r = Pattern.compile("\\[([^]]+)\\]\\[([^]]+)\\]\\[(\\d+)\\]");
+        Matcher m = r.matcher(compositeId);
+        if (m.find()) {
+            String nodeId = m.group(1);
+            String indexName = m.group(2);
+            int shardId = Integer.parseInt(m.group(3));
+            String cluster = null;
+            if (indexName.contains(":")) {
+                // index names and cluster names cannot contain a ':', so this split should be accurate
+                String[] tokens = indexName.split(":", 2);
+                cluster = tokens[0];
+                indexName = tokens[1];
+            }
+            return new ShardProfileId(nodeId, indexName, shardId, cluster);
+        } else {
+            logger.warn("Unable to match input against expected pattern of [nodeId][indexName][shardId]. Input: {}", compositeId);
```

Comment:
Could you add an assertion here so this is caught if it fails during testing? We should never enter this `else` block.

Reply:
Good idea. I need to get used to adding asserts!